File: src/main/java/com/github/mori01231/lifecore/LifeCore.java
package com.github.mori01231.lifecore;

import com.github.mori01231.lifecore.command.DebtCommand;
import com.github.mori01231.lifecore.command.DebugVoteCommand;
import com.github.mori01231.lifecore.command.DropNotifyCommand;
import com.github.mori01231.lifecore.command.DropProtectCommand;
import com.github.mori01231.lifecore.command.EventCommand;
import com.github.mori01231.lifecore.command.GuideCommand;
import com.github.mori01231.lifecore.command.HelpCommand;
import com.github.mori01231.lifecore.command.KiaiCommand;
import com.github.mori01231.lifecore.command.KillNonAdminCommand;
import com.github.mori01231.lifecore.command.LifeCommand;
import com.github.mori01231.lifecore.command.MMIDCommand;
import com.github.mori01231.lifecore.command.NgWordCommand;
import com.github.mori01231.lifecore.command.NoobCommand;
import com.github.mori01231.lifecore.command.PetClickCommand;
import com.github.mori01231.lifecore.command.Pve0Command;
import com.github.mori01231.lifecore.command.Pve1Command;
import com.github.mori01231.lifecore.command.Pve2Command;
import com.github.mori01231.lifecore.command.Pve3Command;
import com.github.mori01231.lifecore.command.PveCommand;
import com.github.mori01231.lifecore.command.RankCommand;
import com.github.mori01231.lifecore.command.ResourceCommand;
import com.github.mori01231.lifecore.command.TrashCommand;
import com.github.mori01231.lifecore.command.TravelCommand;
import com.github.mori01231.lifecore.command.TutorialCommand;
import com.github.mori01231.lifecore.command.VoteCommand;
import com.github.mori01231.lifecore.command.WebsiteCommand;
import com.github.mori01231.lifecore.command.WikiCommand;
import com.github.mori01231.lifecore.config.DropNotifyFile;
import com.github.mori01231.lifecore.config.DropProtectFile;
import com.github.mori01231.lifecore.config.PetClickFile;
import com.github.mori01231.lifecore.config.VotesFile;
import com.github.mori01231.lifecore.listener.CancelJoinAfterStartupListener;
import com.github.mori01231.lifecore.listener.CancelPetClickListener;
import com.github.mori01231.lifecore.listener.CreatureSpawnEventListener;
import com.github.mori01231.lifecore.listener.DeathLoopListener;
import com.github.mori01231.lifecore.listener.DestroyExperienceOrbListener;
import com.github.mori01231.lifecore.listener.DropProtectListener;
import com.github.mori01231.lifecore.listener.FilterNgWordsListener;
import com.github.mori01231.lifecore.listener.DropNotifyListener;
import com.github.mori01231.lifecore.listener.item.GlassHammerItemListener;
import com.github.mori01231.lifecore.listener.item.OreOnlyItemListener;
import com.github.mori01231.lifecore.listener.PlayerJoinListener;
import com.github.mori01231.lifecore.listener.TownyOutlawListener;
import com.github.mori01231.lifecore.listener.TrashListener;
import com.github.mori01231.lifecore.listener.UseAdminSwordListener;
import com.github.mori01231.lifecore.listener.VoteListener;
import com.github.mori01231.lifecore.network.PacketHandler;
import com.github.mori01231.lifecore.util.GCListener;
import com.github.mori01231.lifecore.util.NGWordsCache;
import com.github.mori01231.lifecore.util.PlayerUtil;
import com.sun.net.httpserver.HttpServer;
import io.netty.channel.ChannelPipeline;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandExecutor;
import org.bukkit.entity.Player;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class LifeCore extends JavaPlugin {
    private static LifeCore instance;
    private final GCListener gcListener = new GCListener();
    private final NGWordsCache ngWordsCache = new NGWordsCache();
    private final ExecutorService executorService = Executors.newFixedThreadPool(2);
    public final Executor asyncExecutor = r -> Bukkit.getScheduler().runTaskAsynchronously(this, r);
    private DatabaseConfig databaseConfig;
    private HttpServer httpServer;

    public LifeCore(){
        instance = this;
    }

    public static LifeCore getInstance() {
        return instance;
    }

    @Override
    public void onEnable() {
        // Plugin startup logic

        if (getConfig().getBoolean("enable-gc-detector", false)) {
            gcListener.register();
        }

        VotesFile.load(this);
        PetClickFile.load(this);
        DropProtectFile.load(this);
        DropNotifyFile.load(this);

        databaseConfig = new DatabaseConfig(Objects.requireNonNull(getConfig().getConfigurationSection("database"), "database section is missing"));

        if (getConfig().getBoolean("http-server.enabled", false)) {
            String token = getConfig().getString("http-server.token", "this-should-be-a-random-string");
            if (token == null || "this-should-be-a-random-string".equals(token)) {
                getLogger().warning("Token is not set. Please set a random string to http-server.token in config.yml");
            }
            startHttpServer(getConfig().getInt("http-server.port", 8080));
        }

        getLogger().info("LifeCore has been enabled.");

        registerCommand("wiki", new WikiCommand());
        registerCommand("website", new WebsiteCommand());
        registerCommand("help", new HelpCommand());
        registerCommand("guide", new GuideCommand());
        registerCommand("tutorial", new TutorialCommand());
        registerCommand("travel", new TravelCommand());
        registerCommand("pve", new PveCommand());
        registerCommand("pve0", new Pve0Command());
        registerCommand("pve1", new Pve1Command());
        registerCommand("pve2", new Pve2Command());
        registerCommand("pve3", new Pve3Command());
        registerCommand("life", new LifeCommand());
        registerCommand("resource", new ResourceCommand());
        registerCommand("event", new EventCommand());
        registerCommand("rank", new RankCommand());
        registerCommand("trash", new TrashCommand());
        registerCommand("kiai", new KiaiCommand());
        registerCommand("noob", new NoobCommand());
        registerCommand("killnonadmin", new KillNonAdminCommand());
        registerCommand("mmid", new MMIDCommand());
        registerCommand("debt", new DebtCommand());
        registerCommand("vote", new VoteCommand());
        registerCommand("debugvote", new DebugVoteCommand());
        registerCommand("petclick", new PetClickCommand());
        registerCommand("dropprotect", new DropProtectCommand(this));
        registerCommand("ngword", new NgWordCommand(this));
        registerCommand("dropnotify", new DropNotifyCommand());

        this.saveDefaultConfig();

        getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");

        registerEvents();

        try {
            DBConnector.init();
        } catch (Exception e) {
            getLogger().severe("Failed to connect to database.");
            e.printStackTrace();
        }

        // register channel handler
        for (Player player : Bukkit.getOnlinePlayers()) {
            try {
                ngWordsCache.loadAsync(player.getUniqueId());
                PlayerUtil.getChannel(player).pipeline()
                        .addBefore("packet_handler", "lifecore", new PacketHandler(player));
            } catch (Exception e) {
                getSLF4JLogger().warn("Failed to inject channel handler to player " + player.getName(), e);
            }
        }

        // check for negative max health
        Bukkit.getScheduler().runTaskTimer(this, () -> {
            for (Player player : Bukkit.getOnlinePlayers()) {
                DeathLoopListener.checkAttribute(player);
            }
        }, 100, 100);
    }

    private void startHttpServer(int port) {
        try {
            httpServer = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        httpServer.createContext("/accept-rules", new RequestHandler(this));
        httpServer.setExecutor(executorService);
        httpServer.start();
    }
    
    private void registerCommand(@NotNull String name, @NotNull CommandExecutor executor) {
        Objects.requireNonNull(getCommand(name), name + " is not registered in plugin.yml").setExecutor(executor);
    }

    @Override
    public void onDisable() {
        // Plugin shutdown logic

        VotesFile.save(this);
        PetClickFile.save(this);
        DropProtectFile.save(this);
        DropNotifyFile.save(this);
        DBConnector.close();
        executorService.shutdownNow();
        httpServer.stop(1);
        gcListener.unregister();

        // unregister all channel handlers
        for (Player player : Bukkit.getOnlinePlayers()) {
            ChannelPipeline pipeline = PlayerUtil.getChannel(player).pipeline();
            try {
                if (pipeline.get("lifecore") != null) {
                    pipeline.remove("lifecore");
                }
            } catch (NoSuchElementException ignored) {}
        }

        getLogger().info("LifeCore has been disabled.");
    }

    public void registerEvents() {
        PluginManager pm = getServer().getPluginManager();

        pm.registerEvents(new CreatureSpawnEventListener(), this);
        pm.registerEvents(new TrashListener(), this);
        pm.registerEvents(new UseAdminSwordListener(), this);
        pm.registerEvents(new PlayerJoinListener(this), this);
        pm.registerEvents(new CancelJoinAfterStartupListener(), this);
        pm.registerEvents(new DeathLoopListener(), this);
        pm.registerEvents(new DropProtectListener(), this);
        pm.registerEvents(new DropNotifyListener(), this);

        // Items
        pm.registerEvents(new OreOnlyItemListener(), this);
        pm.registerEvents(new GlassHammerItemListener(), this);

        if (getConfig().getBoolean("destroy-experience-orb-on-chunk-load", false)) {
            pm.registerEvents(new DestroyExperienceOrbListener(), this);
        }

        try {
            Class.forName("de.Keyle.MyPet.MyPetApi");
            pm.registerEvents(new CancelPetClickListener(), this);
        } catch (Exception | NoClassDefFoundError e) {
            getSLF4JLogger().warn("MyPet not detected, skipping event listener registration");
        }

        try {
            Class.forName("com.vexsoftware.votifier.model.VotifierEvent");
            pm.registerEvents(new VoteListener(), this);
        } catch (Exception | NoClassDefFoundError e) {
            getSLF4JLogger().warn("Votifier not detected, skipping event listener registration");
        }

        try {
            Class.forName("com.palmergames.bukkit.towny.TownyAPI");
            pm.registerEvents(new TownyOutlawListener(), this);
        } catch (Exception | NoClassDefFoundError e) {
            getSLF4JLogger().warn("Towny not detected, skipping event listener registration");
        }

        try {
            Class.forName("net.azisaba.ryuzupluginchat.event.AsyncGlobalMessageEvent");
            pm.registerEvents(new FilterNgWordsListener(this), this);
        } catch (Exception | NoClassDefFoundError e) {
            getSLF4JLogger().warn("RyuZUPluginChat not detected, skipping event listener registration");
        }
    }

    @NotNull
    public DatabaseConfig getDatabaseConfig() {
        return Objects.requireNonNull(databaseConfig, "databaseConfig is null");
    }

    @NotNull
    public GCListener getGcListener() {
        return gcListener;
    }

    public @NotNull NGWordsCache getNgWordsCache() {
        return ngWordsCache;
    }
}

File: src/main/java/com/github/mori01231/lifecore/LifeCore.kt
package com.github.mori01231.lifecore

import com.github.mori01231.lifecore.block.CustomBlockManager
import com.github.mori01231.lifecore.command.*
import com.github.mori01231.lifecore.config.*
import com.github.mori01231.lifecore.data.DataLoader
import com.github.mori01231.lifecore.gui.CommandListScreen
import com.github.mori01231.lifecore.gui.DropProtectScreen
import com.github.mori01231.lifecore.gui.MapListScreen
import com.github.mori01231.lifecore.gui.TrashProtectScreen
import com.github.mori01231.lifecore.listener.*
import com.github.mori01231.lifecore.listener.item.*
import com.github.mori01231.lifecore.network.PacketHandler
import com.github.mori01231.lifecore.region.PlayerRegionManager
import com.github.mori01231.lifecore.util.*
import com.github.mori01231.lifecore.util.YAML.yaml
import com.sun.net.httpserver.HttpServer
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import org.bukkit.Bukkit
import org.bukkit.command.CommandExecutor
import org.bukkit.entity.ItemFrame
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.net.InetSocketAddress
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import kotlin.collections.component1
import kotlin.collections.component2

class LifeCore : JavaPlugin() {
    val gcListener = GCListener(this)
    @JvmField
    val ngWordsCache = NGWordsCache()
    private val executorService = Executors.newFixedThreadPool(2)
    @JvmField
    val asyncExecutor = Executor { Bukkit.getScheduler().runTaskAsynchronously(this, it) }
    val lifeCoreConfig = LifeCoreConfig.load(this)
    val playerRegionManager = PlayerRegionManager()
    val dataLoader = DataLoader(slF4JLogger, getAsFileSystem().getPath("/data"))
    private var databaseConfig: DatabaseConfig? = null
    lateinit var voteConfig: VoteConfig
        private set
    private var httpServer: HttpServer? = null
    lateinit var dropProtectConfig: DropProtectConfig
        private set
    lateinit var trashProtectConfig: TrashProtectConfig
        private set
    lateinit var mythicItemProtectConfig: MythicItemProtectConfig
        private set
    val customBlockManager = CustomBlockManager(this)
    var customModelDataWrench = 0
    var customModelDataBlank = 0
    var customModelDataBlankShapeDownLeft = 0
    var customModelDataBlankShapeDownRight = 0

    init {
        instance = this

        preloadClasses()
    }

    override fun onEnable() {
        // Plugin startup logic
        customModelDataWrench = config.getInt("custom-model-data.wrench", 0)
        customModelDataBlank = config.getInt("custom-model-data.blank", 0)
        customModelDataBlankShapeDownLeft = config.getInt("custom-model-data.blank-shape-down-left", 0)
        customModelDataBlankShapeDownRight = config.getInt("custom-model-data.blank-shape-down-right", 0)
        customBlockManager.init()

        if (config.getBoolean("enable-gc-detector", false)) {
            gcListener.register()
        }
        VotesFile.load(this)
        PetClickFile.load(this)
        DamageLogFile.load(this)
        DropNotifyFile.load(this)
        dataFolder.resolve("drop-protect.yml").let {
            if (!it.exists()) {
                it.writeText(yaml.encodeToString(DropProtectConfig()))
            }
            dropProtectConfig = yaml.decodeFromString(it.readText())
        }
        dataFolder.resolve("trash-protect.yml").let {
            if (!it.exists()) {
                it.writeText(yaml.encodeToString(TrashProtectConfig()))
            }
            trashProtectConfig = yaml.decodeFromString(it.readText())
        }
        dataFolder.resolve("mythic-item-protect.yml").let {
            if (!it.exists()) {
                it.writeText(yaml.encodeToString(MythicItemProtectConfig()))
            }
            mythicItemProtectConfig = yaml.decodeFromString(it.readText())
        }

        // save every 5 minutes
        Bukkit.getScheduler().runTaskTimerAsynchronously(this, 20 * 60 * 5, 20 * 60 * 5) {
            dataFolder.resolve("drop-protect.yml").writeText(dropProtectConfig.encode())
            dataFolder.resolve("trash-protect.yml").writeText(trashProtectConfig.encode())
            dataFolder.resolve("mythic-item-protect.yml").writeText(mythicItemProtectConfig.encode())
        }

        // save every 10 seconds
        Bukkit.getScheduler().runTaskTimerAsynchronously(this, 20 * 10, 20 * 10) {
            customBlockManager.saveAll()
        }

        databaseConfig = config.getYamlMap("database").decode(DatabaseConfig.serializer())
        dataFolder.resolve("config-vote.yml").let {
            if (!it.exists()) {
                it.writeText(yaml.encodeToString(VoteConfig()))
            }
            voteConfig = yaml.decodeFromString(it.readText())
        }

        val httpServerConfig =
            config.getConfigurationSection("http-server")
                ?.toReadonlyNode()
                ?.decode(HttpServerConfig.serializer())
        if (httpServerConfig != null && httpServerConfig.enabled) {
            if (httpServerConfig.token == HttpServerConfig.DEFAULT_TOKEN) {
                logger.warning("Token is not set. Please set a random string to http-server.token in config.yml")
            }
            startHttpServer(httpServerConfig.port)
        }

        registerCommand("wiki", WikiCommand(this))
        registerCommand("website", WebsiteCommand(this))
        registerCommand("help", HelpCommand(this))
        registerCommand("guide", GuideCommand(this))
        registerCommand("tutorial", TutorialCommand(this))
        registerCommand("travel", TransferCommand(this, "lifetravel"))
        registerCommand("pve", PveCommand(this))
        registerCommand("pve0", TransferCommand(this, "lifepve"))
        registerCommand("pve1", TransferCommand(this, "lifepve1"))
        registerCommand("pve2", TransferCommand(this, "lifepve2"))
        registerCommand("pve3", TransferCommand(this, "lifepve3"))
        registerCommand("life", TransferCommand(this, "life"))
        registerCommand("resource", TransferCommand(this, "liferesource"))
        registerCommand("event", TransferCommand(this, "lifeevent"))
        registerCommand("townserver", TownServerCommand(this))
        registerCommand("rank", RankCommand(this))
        registerCommand("trash", TrashCommand())
        registerCommand("kiai", KiaiCommand())
        registerCommand("noob", NoobCommand())
        registerCommand("killnonadmin", KillNonAdminCommand())
        registerCommand("mmid", MMIDCommand())
        registerCommand("debt", DebtCommand(this))
        registerCommand("vote", VoteCommand(this))
        registerCommand("debugvote", DebugVoteCommand(this))
        registerCommand("petclick", PetClickCommand())
        registerCommand("ngword", NgWordCommand(this))
        registerCommand("dropnotify", DropNotifyCommand())
        registerCommand("dropprotect", DropProtectCommand(this))
        registerCommand("trashprotect", TrashProtectCommand(this))
        registerCommand("protect", ProtectCommand(this))
        registerCommand("damagelog", DamageLogCommand())
        registerCommand("servermoney", ServerMoneyCommand(this))
        registerCommand("fixtime", FixTimeCommand)
        registerCommand("lifecoreconfig", LifeCoreConfigCommand(this))
        registerCommand("townconfig", TownConfigCommand(this))
        registerCommand("schedulerestart", ScheduleRestartCommand(this))
        registerCommand("gclistenerrestartextendtimecommand", GCListenerRestartExtendTimeCommand(this))
        registerCommand("lifecoreutil", LifeCoreUtilCommand(this))
        registerCommand("commandlist", CommandListCommand)
        registerCommand("maplist", MapListCommand)
        registerCommand("worldcreate", WorldCreateCommand())
        registerCommand("respawn") { _, _, _, args ->
            args.getOrNull(0)?.let { Bukkit.getPlayerExact(it)?.spigot()?.respawn() }
            true
        }
        saveDefaultConfig()
        server.messenger.registerOutgoingPluginChannel(this, "BungeeCord")
        registerEvents()
        try {
            DBConnector.init(this)
        } catch (e: Exception) {
            logger.severe("Failed to connect to database.")
            e.printStackTrace()
        }

        // register channel handler
        for (player in Bukkit.getOnlinePlayers()) {
            try {
                ngWordsCache.loadAsync(player.uniqueId)
                PlayerUtil.getChannel(player).pipeline()
                    .addBefore("packet_handler", "lifecore", PacketHandler(player))
            } catch (e: Exception) {
                slF4JLogger.warn("Failed to inject channel handler to player " + player.name, e)
            }
        }

        // check for negative max health
        Bukkit.getScheduler().runTaskTimer(this, Runnable {
            for (player in Bukkit.getOnlinePlayers()) {
                DeathLoopListener.checkAttribute(this, player)
            }
        }, 100, 100)

        // check for maps
        Bukkit.getScheduler().runTaskTimer(this, Runnable {
            Bukkit.getOnlinePlayers().forEach player@ { player ->
                player.inventory.contents.forEach { item ->
                    @Suppress("SENSELESS_COMPARISON")
                    if (item != null) {
                        MapUtil.initializeMapRenderer(player, item)
                    }
                }
            }
            if (Bukkit.getOnlinePlayers().isEmpty()) return@Runnable
            Bukkit.getWorlds().forEach world@ { world ->
                val list = world.getEntitiesByClass(ItemFrame::class.java)
                Bukkit.getScheduler().runTaskAsynchronously(this, Runnable {
                    list.forEach { itemFrame ->
                        val item = itemFrame.item
                        Bukkit.getOnlinePlayers().forEach { player ->
                            MapUtil.initializeMapRenderer(player, item)
                        }
                    }
                })
            }
        }, 20 * 30, 20 * 30)

        Bukkit.getScheduler().runTaskTimer(this, Runnable {
            customBlockManager.getLoadedStates().forEach { (location, state) ->
                state.getBlock().tick(customBlockManager, location, state)?.let {
                    customBlockManager.setState(location.toBukkitLocation(), it)
                }
            }
        }, 1, 1)

        logger.info("LifeCore has been enabled.")
    }

    private fun startHttpServer(port: Int) {
        httpServer = try {
            HttpServer.create(InetSocketAddress("0.0.0.0", port), 0).apply {
                createContext("/accept-rules", RequestHandler(this@LifeCore))
                executor = executorService
                start()
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private fun registerCommand(name: String, executor: CommandExecutor) {
        getCommand(name).nonNull("$name is not registered in plugin.yml").setExecutor(executor)
    }

    override fun onDisable() {
        // Plugin shutdown logic
        runCatching { VotesFile.save(this) }
        runCatching { PetClickFile.save(this) }
        runCatching { DamageLogFile.save(this) }
        runCatching { DropNotifyFile.save(this) }
        runCatching { dataFolder.resolve("drop-protect.yml").writeText(dropProtectConfig.encode()) }
        runCatching { DBConnector.close() }
        runCatching { executorService.shutdownNow() }
        runCatching { httpServer?.stop(1) }
        runCatching { gcListener.unregister() }
        runCatching { customBlockManager.saveAll() }

        runCatching {
            // unregister all channel handlers
            for (player in Bukkit.getOnlinePlayers()) {
                val pipeline = PlayerUtil.getChannel(player).pipeline()
                try {
                    if (pipeline["lifecore"] != null) {
                        pipeline.remove("lifecore")
                    }
                } catch (_: NoSuchElementException) {
                }
            }
        }
        logger.info("LifeCore has been disabled.")
    }

    private fun preloadClasses() {
        for (i in 0..10) {
            preloadClass("com.github.mori01231.lifecore.LifeCore\$onDisable\$$i", false)
        }
        preloadClass("com.github.mori01231.lifecore.lib.com.charleskorn.kaml.Yaml\$encodeToString\$writer\$1")
        preloadClass("com.github.mori01231.lifecore.lib.com.charleskorn.kaml.YamlOutput")
        preloadClass("com.github.mori01231.lifecore.lib.org.yaml.snakeyaml.Yaml")
        preloadClass("com.github.mori01231.lifecore.lib.org.yaml.snakeyaml.nodes.CollectionNode")
        preloadClass("com.github.mori01231.lifecore.lib.org.yaml.snakeyaml.nodes.SequenceNode")
        preloadClass("com.github.mori01231.lifecore.lib.org.snakeyaml.engine.v2.api.DumpSettings")
        preloadClass("com.github.mori01231.lifecore.lib.org.snakeyaml.engine.v2.api.DumpSettingsBuilder")
        preloadClass("com.github.mori01231.lifecore.lib.org.yaml.snakeyaml.nodes.MappingNode")
        preloadClass("com.github.mori01231.lifecore.lib.org.snakeyaml.engine.v2.serializer.NumberAnchorGenerator")
        preloadClass("com.github.mori01231.lifecore.lib.org.snakeyaml.engine.v2.common.NonPrintableStyle")
        preloadClass("com.github.mori01231.lifecore.lib.org.snakeyaml.engine.v2.emitter.Emitter")
        preloadClass("com.github.mori01231.lifecore.lib.org.yaml.snakeyaml.constructor.ConstructorException")
    }

    private fun preloadClass(name: String, required: Boolean = true) {
        try {
            Class.forName(name)
        } catch (e: ClassNotFoundException) {
            if (required) {
                throw e
            }
        }
    }

    private fun runCatching(action: () -> Unit) {
        try {
            action()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun registerEvents() {
        val pm = server.pluginManager
        pm.registerEvents(CreatureSpawnEventListener(this), this)
        pm.registerEvents(TrashListener(this), this)
        pm.registerEvents(UseAdminSwordListener(this), this)
        pm.registerEvents(PlayerJoinListener(this), this)
        pm.registerEvents(CancelJoinAfterStartupListener(), this)
        pm.registerEvents(DeathLoopListener(), this)
        pm.registerEvents(DropNotifyListener(), this)
        pm.registerEvents(DropProtectScreen.EventListener(this), this)
        pm.registerEvents(TrashProtectScreen.EventListener(this), this)
        pm.registerEvents(UnableCraftListener(), this)
        pm.registerEvents(AZISAVIORListener(this), this)
        pm.registerEvents(DamageLogListener(), this)
        pm.registerEvents(UnusableDyeListener(), this)
        pm.registerEvents(WandItemListener(this), this)
        pm.registerEvents(EscapeLobbyListener(this), this)
        pm.registerEvents(PreventOpenMerchantMenuListener(), this)
        pm.registerEvents(PreventStorageBoxSellListener(this), this)
        pm.registerEvents(CancelMythicItemPlaceListener(), this)
        pm.registerEvents(GameModeChangeLoggerListener(this), this)
        pm.registerEvents(FixMythicItemListener, this)
        pm.registerEvents(VoidListener, this)
        pm.registerEvents(Dice1ItemListener(this), this)
        pm.registerEvents(OverrideTabCompleteListener(), this)
        pm.registerEvents(KotlinLoveItemListener(this), this)
        pm.registerEvents(CommandListScreen.EventListener(), this)
        pm.registerEvents(UpdateInventoryOnCloseListener(this), this)
        pm.registerEvents(PromptSignListener, this)
        pm.registerEvents(PicksawItemListener(dataLoader), this)
        pm.registerEvents(BlockListener, this)
        pm.registerEvents(MapListScreen.EventListener, this)
        pm.registerEvents(PlayerQuitListener, this)
        pm.registerEvents(WorkaroundLuckPerms, this)

        // Items
        pm.registerEvents(OreOnlyItemListener(), this)
        pm.registerEvents(GlassHammerItemListener(), this)
        pm.registerEvents(FlashLightListener(), this)
        pm.registerEvents(LavaSpongeItemListener(), this)
        if (config.getBoolean("destroy-experience-orb-on-chunk-load", false)) {
            pm.registerEvents(DestroyExperienceOrbListener(), this)
        }
        try {
            Class.forName("de.Keyle.MyPet.MyPetApi")
            pm.registerEvents(CancelPetClickListener(), this)
        } catch (e: ClassNotFoundException) {
            slF4JLogger.warn("MyPet not detected, skipping event listener registration")
        }
        try {
            Class.forName("com.vexsoftware.votifier.model.VotifierEvent")
            pm.registerEvents(VoteListener(this), this)
        } catch (e: ClassNotFoundException) {
            slF4JLogger.warn("Votifier not detected, skipping event listener registration")
        }
        try {
            Class.forName("com.palmergames.bukkit.towny.TownyAPI")
            pm.registerEvents(TownyOutlawListener(), this)
            TownSpecificListener(this)
                .apply { startTask() }
                .apply { pm.registerEvents(this, this@LifeCore) }
        } catch (e: ClassNotFoundException) {
            slF4JLogger.warn("Towny not detected, skipping event listener registration")
            e.printStackTrace()
        }
        try {
            Class.forName("net.azisaba.ryuzupluginchat.event.AsyncGlobalMessageEvent")
            pm.registerEvents(FilterNgWordsListener(this), this)
        } catch (e: ClassNotFoundException) {
            slF4JLogger.warn("RyuZUPluginChat not detected, skipping event listener registration")
        }
        try {
            Class.forName("net.azisaba.rarity.api.RarityAPI")
            Class.forName("net.azisaba.itemstash.ItemStash")
            pm.registerEvents(DropProtectListener(this), this)
        } catch (e: ClassNotFoundException) {
            slF4JLogger.warn("Rarity and/or ItemStash not detected, skipping event listener registration")
        }
    }

    fun getDatabaseConfig(): DatabaseConfig {
        return databaseConfig.nonNull("databaseConfig is null")
    }

    companion object {
        @get:Deprecated("Using this field can cause a memory leak", ReplaceWith("getPlugin(LifeCore::class.java)"))
        @JvmStatic
        lateinit var instance: LifeCore
            private set
    }

    fun getAsFileSystem() =
        this::class.java
            .protectionDomain
            .codeSource
            .location
            .file
            .let { FileSystemUtil.openFileAsFileSystem(File(it)) }
}
